components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java

```diff
@@ -1783,134 +1783,362 @@ private int handleStartIndexEqualsNULL(Integer startIndex) {
                     }
                 } else {
                     if (log.isDebugEnabled()) {
-                        log.debug("SCIM is disabled for the user-store domain : " + userStoreDomainName + ". Hence " +
-                                "group with name : " + roleName + " is excluded in the response.");
+                        log.debug("SCIM is disabled for the user-store domain : " + userStoreDomainName + ". Hence "
+                                + "group with name : " + roleName + " is excluded in the response.");
                     }
                 }
             }
         } catch (org.wso2.carbon.user.core.UserStoreException e) {
             String errMsg = "Error in obtaining role names from user store.";
             errMsg += e.getMessage();
             throw new CharonException(errMsg, e);
+        } catch (org.wso2.carbon.user.api.UserStoreException e) {
+            String errMsg = "Error in retrieving role names from user store.";
+            throw new CharonException(errMsg, e);
         } catch (IdentitySCIMException | BadRequestException e) {
             throw new CharonException("Error in retrieving SCIM Group information from database.", e);
         }
         //set the totalResults value in index 0
-        groupList.set(0, groupList.size()-1);
+        groupList.set(0, groupList.size() - 1);
         return groupList;
     }
 
+    private Set<String> getRoleNamesForGroupsEndpoint(String domainName)
+            throws UserStoreException, IdentitySCIMException {
+
+        SCIMGroupHandler groupHandler = new SCIMGroupHandler(carbonUM.getTenantId());
+        if (StringUtils.isEmpty(domainName)) {
+            return groupHandler.listSCIMRoles();
+        } else {
+            // If the domain is not specified create a filter with SW operator and retrieve the role names by
+            // appending the domain name.
+            String attributeValue = domainName + CarbonConstants.DOMAIN_SEPARATOR;
+            List<String> roleList = Arrays.asList(getRoleNames(SCIMCommonConstants.SW, attributeValue));
+            Set<String> roleNames = new HashSet<>(roleList);
+            roleNames.addAll(roleList);
+            return roleNames;
+        }
+    }
 
-    private List<Object> filterGroups(Node node, Map<String, Boolean> requiredAttributes)
+    /**
+     * @param rootNode
+     * @param domainName
+     * @param requiredAttributes
+     * @return
+     * @throws NotImplementedException
+     * @throws CharonException
+     */
+    private List<Object> filterGroups(Node rootNode, int startIndex, Integer count, String sortBy, String sortOrder,
+            String domainName, Map<String, Boolean> requiredAttributes)
             throws NotImplementedException, CharonException {
 
-        if(node.getLeftNode() != null || node.getRightNode() != null){
+        // Handle count equals NULL scenario.
+        count = handleLimitEqualsNULL(count);
+        if (rootNode instanceof ExpressionNode) {
```

Comment:
Even though it is expression node, we only support four filters, how do we handle unsupported filters?

Reply:
Fixed: e0b27781f03c5dc319612f3d1924e70dce6c2f21